### ISSUE TITLE
Could you provide the attention-map visualization code ?

### ISSUE THREAD

**Issue by Yang-bug-star** [Pter61/context-i2w#11]

**Comment by Pter61** [comment]
Of course, here is the relevant code, hope it can help you:

`def get_attn_map(root, imagepath_attn_w, save_path):
    for bath_idx in tqdm(range(len(imagepath_attn_w))):

        attn_map_all = imagepath_attn_w[bath_idx][1]
        for layer_idx in range(len(attn_map_all)):
            for sample_idx in tqdm(range(len(imagepath_attn_w[bath_idx][0]))):
                imgname, attn_map, tgt_imgname, text = imagepath_attn_w[bath_idx][0][sample_idx], attn_map_all[layer_idx][sample_idx], imagepath_attn_w[bath_idx][2][sample_idx], imagepath_attn_w[bath_idx][3][sample_idx]
                save_detail = "\"reference\": \"%s\", \"target_hard\": \"%s\"\n%s" % (
                imgname[:-4], tgt_imgname[:-4], text)
                save_path_root = os.path.join(save_path, "%s_%s" % (imgname[:-4], tgt_imgname[:-4]))
                check_folder_exist(save_path_root)
                shutil.copyfile(os.path.join(root, imgname), os.path.join(save_path_root, imgname))
                save_path_layer = os.path.join(save_path_root, "layer_%d" % layer_idx)
                check_folder_exist(save_path_layer)
                attn_map = torch.sum(attn_map, dim=0)

                for head_idx in range(attn_map.shape[0]):
                    if attn_map.shape[0] > 1:
                        save_path_head = os.path.join(save_path_layer, "head_%d" % head_idx)
                        check_folder_exist(save_path_head)
                    else:
                        save_path_head = save_path_layer
                    attn_map_head = attn_map[head_idx].squeeze(0)
                    attn_map_head = attn_map_head.cpu().numpy()

                    for query_idx in range(attn_map_head.shape[0]):
                        save_path_query = save_path_head
                        if attn_map.shape[0] > 1:
                            save_path_query = os.path.join(save_path_query, "layer_%d-head_%d-%s_%d.jpg" % (layer_idx, head_idx, imgname[:-4], query_idx))
                        else:
                            save_path_query = os.path.join(save_path_query, "layer_%d-sum-%s_%d.jpg" % (
                            layer_idx, imgname[:-4], query_idx))
                        if os.path.exists(save_path_query):
                            continue
                        # Visualize attention map
                        attn_visualiztion(root, imgname, attn_map_head[query_idx][:-4], save_path_query, grid_size=16)
                        # Write caption of this sample
                        caption_file = os.path.join(save_path_root, "caption.txt")
                        with open(caption_file, "w") as f:
                            f.write(f"%s" % save_detail)
                        f.close()
    print("Done! Saved to", save_path)`

If you find it useful, feel free to star my repo~

**Comment by Yang-bug-star** [comment]
Hello, could you share attn_visualiztion function details ? And I'm curious about which parts imagepath_attn_w correspond to in the code.

**Comment by Pter61** [comment]
The imagepath_attn_w is the attention map of each learnable query from the last block, which is mentioned in our paper.

Here is the sample code:
```
        attn_weight = []
        # blocks
        latents = repeat(self.latents, "n d -> b T n d", b=b, T=T)
        for attn, ff in self.layers:
            # latents = attn(x, latents) + latents

            latents_tmp, attn_weight_tmp = attn(x, latents)
            latents = latents_tmp + latents
            latents = ff(latents) + latents
            # if is the last layer
            if attn == self.layers[-1][0]:
                attn_weight.append(attn_weight_tmp)
```
